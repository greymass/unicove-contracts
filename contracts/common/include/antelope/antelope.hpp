#pragma once

#include <eosio.system/eosio.system.hpp>

using namespace eosio;
using namespace std;

namespace antelope {

static inline symbol default_system_token_symbol = symbol("UNKNOWN", 0);
static inline asset  default_system_token_asset  = asset(0, default_system_token_symbol);

struct token_definition
{
   optional<checksum256> chain;
   name                  contract;
   symbol                symbol;
};

struct token_distribution
{
   asset circulating = default_system_token_asset;
   asset locked      = default_system_token_asset;
   asset max         = default_system_token_asset;
   asset staked      = default_system_token_asset;
   asset supply      = default_system_token_asset;
};

struct token
{
   token_definition             id;
   optional<token_distribution> distribution;
};

struct token_balance
{
   token token;
   asset balance;
};

} // namespace antelope
