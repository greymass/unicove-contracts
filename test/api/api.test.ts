import {beforeEach, describe, expect, test} from 'bun:test'

import {apiContract, contracts, resetContracts} from '../helpers'
import {Asset, Name} from '@wharfkit/antelope'

describe(apiContract, () => {
    beforeEach(async () => {
        await resetContracts()
    })

    describe('action::setconfig', () => {
        describe('success', () => {
            test('set config', async () => {
                await contracts.api.actions
                    .setconfig(['foo', 'foo.msig', 'foo.token', '4,FOO'])
                    .send()
                const rows = await contracts.api.tables.config().getTableRows()
                expect(rows).toHaveLength(1)
                expect(Name.from(rows[0].system_contract).equals('foo')).toBeTrue()
                expect(Name.from(rows[0].system_contract_msig).equals('foo.msig')).toBeTrue()
                expect(Name.from(rows[0].system_token_contract).equals('foo.token')).toBeTrue()
                expect(Asset.Symbol.from(rows[0].system_token_symbol).equals('4,FOO')).toBeTrue()
            })
            test('defaults', async () => {
                const rows = await contracts.api.tables.config().getTableRows()
                expect(rows).toHaveLength(1)
                expect(Name.from(rows[0].system_contract).equals('eosio')).toBeTrue()
                expect(Name.from(rows[0].system_contract_msig).equals('eosio.msig')).toBeTrue()
                expect(Name.from(rows[0].system_token_contract).equals('eosio.token')).toBeTrue()
                expect(Asset.Symbol.from(rows[0].system_token_symbol).equals('4,EOS')).toBeTrue()
            })
        })
        describe('error', () => {
            test('require contract auth', async () => {
                const action = contracts.api.actions
                    .setconfig(['eosio', 'eosio.msig', 'eosio.token', '4,EOS'])
                    .send('alice')
                expect(action).rejects.toThrow('missing required authority api')
            })
        })
    })
})
