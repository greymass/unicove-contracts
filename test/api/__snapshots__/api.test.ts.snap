// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`api action::addtoken success 1`] = `
{
  "contract": "eosio.token",
  "id": 0,
  "symbol": "4,EOS",
}
`;

exports[`api action::addtoken success 2`] = `
{
  "contract": "scrap",
  "id": 1,
  "symbol": "0,SCRAP",
}
`;

exports[`api action::removetoken success 1`] = `
{
  "contract": "eosio.token",
  "id": 0,
  "symbol": "4,EOS",
}
`;

exports[`api action::removetoken success 2`] = `
{
  "contract": "scrap",
  "id": 1,
  "symbol": "0,SCRAP",
}
`;

exports[`api action::removetoken success 3`] = `
{
  "contract": "eosio.token",
  "id": 0,
  "symbol": "4,EOS",
}
`;

exports[`api action::addtoken success add tokens 1`] = `
{
  "contract": "eosio.token",
  "id": 0,
  "symbol": "4,EOS",
}
`;

exports[`api action::addtoken success add tokens 2`] = `
{
  "contract": "scrap",
  "id": 1,
  "symbol": "0,SCRAP",
}
`;

exports[`api action::removetoken success remove tokens 1`] = `
{
  "contract": "eosio.token",
  "id": 0,
  "symbol": "4,EOS",
}
`;

exports[`api action::removetoken success remove tokens 2`] = `
{
  "contract": "scrap",
  "id": 1,
  "symbol": "0,SCRAP",
}
`;

exports[`api action::removetoken success remove tokens 3`] = `
{
  "contract": "eosio.token",
  "id": 0,
  "symbol": "4,EOS",
}
`;
